share/doc/wake/logging.md

```diff
@@ -0,0 +1,64 @@
+# Logging
+
+This article gives a high level overview of how wake logging works.
+Check out [print.wake](https://github.com/sifive/wake/blob/master/share/wake/lib/core/print.wake)
+for commented code.
+
+## How Logging works on Wake Execution commands
+
+Every `Job` will emit a stderr and a stdout.
+
+When writing your wake code, you can control where each goes with `setPlanStderr` and `setPlanStdout` before you `runJob`.
+We call this which `stream` each goes to (we often refer to a `stream` as a `logLevel`).
+
+Regardless of which `stream` you direct the `Job`'s `Stderr` and `Stdout`,
+`wake` will store all the text in its database,
+discarding the information about the `stream` it was destined for at the time it was stored.
+It will however know whether it came from the `Job`'s stderr/stdout.
+
+So what's the difference between the different streams?
+What you see on your console when you run wake.
+It controls *what* you see (depending on the command line flags) and what it looks like (what color it is).
+It also controls which *actual* stream it is going to coming out of `wake`
+(stderr or stdout, or other file descriptors 3-5).
+
+There are a number of built in streams in wake which all start with `log...`.
+You can also make your own stream using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as,
+or (more commonly) to grab output from a `Job` for passing to another `Job`,
+consuming within your wake program, or redirecting outside of wake using shell redirect.
+
+Unless overridden with `--fd:2=...` the *only* thing that ever appears on stderr coming out of wake is `logError`,
+and it always appears there.
+
+For stdout, the following table should be read top to bottom considering the command line arguments you've passed to wake.
+If you match a line in the table, stop!
+This is what will appear on the console as you watch wake execute, and what is sent to wake's stdout:
+
+|Flag               | logDebug | logInfo | logEcho | logReport | logWarning | logError | logNever | mkLogLevel "foo" Green|
+|-------------------|----------|---------|---------|-----------|------------|----------|----------|----------------------
+|Color on Console   |  Blue    |  Dim    |  default| Magenta   |  Yellow    |   Red    |          |   Green         |          
```

Comment:
Your response is contradictory, since you say the "console default" and then you say whatever scalatest overrides the console default with. Your console default is determined by your terminal emulator's settings, specifically the foreground color (normally you set both the foreground and the background, since most of the time you either want dark text on a light background or light text on a dark background). Programs can change the foreground and background colors by setting control codes that affect either the foreground or background colors, and the colors used by wake appear to be the [ANSI 3-bit/4-bit colors](https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit) ((black + white + 6 other colors) &times; bright modifier).

Most well-behaved programs will not emit color codes to stdout/stderr when not connected to a tty (e.g. stdout is piped to a file or another program), but I've noticed that most Scala programs really do not seem to follow those conventions at all. I have no idea what happens in wake when a program prints color codes and whether wake overrides them or whether the program's original color codes come out.

Reply:
the program's original color codes come out. I mean, there are 2 arguments to the color (in wake) -- the color override and the intensity override. "console default" isn't the right term, but I want to say something like "whatever color you would have seen if you were just running it, not using wake". do you have a better suggestion for that term?

Reply:
> (but have not re-read the code in question since you wrote this comment)

Having read the code now, here's an explanation of how it works:

In wake-land, [`makeLogLevel2` eventually makes a call](https://github.com/sifive/wake/blob/e596242b4a009550e2bd06ea21206f9aedeeea18/share/wake/lib/core/print.wake#L61) to the [primitive function `colour`](https://github.com/sifive/wake/blob/a84668a15cdb076e9796631aa32f6cb7f4941a08/src/runtime/string.cpp#L391-L398), which accepts two arguments: a string name (the logger name), and an integer colo<s>u</s>r. Note that this color encodes both the hue and the the intensity using the following bit mask:

```
[MSB]   5    4          3        2  1  0  [LSB]
      [intensity]      ???      [  hue  ]

intensity = 0 => unset
intensity = 1 => dim
intensity = 2 => bright

??? = apparently set to 1 (via adding 8) when you want to set a color, including black,
      but otherwise ignored by the rest of the wake code? Maybe just a sentinel
      so that when you explicitly want to set a color of black, the rest of the wake
      code can see that you explicitly wanted black.

hue = standard 3-bit ANSI encoding
```

The primitive function `colour` just calls out to [`set_status_colour()`](https://github.com/sifive/wake/blob/e596242b4a009550e2bd06ea21206f9aedeeea18/src/runtime/status.cpp#L280), which is a C++ function that just stores the name and integer color (still encoded) into a hash table. Later on, when wake wants to actually write something to a logger/stream/whatever we're calling it, it calls [`status_write()`](https://github.com/sifive/wake/blob/e596242b4a009550e2bd06ea21206f9aedeeea18/src/runtime/status.cpp#L300-L314), which decodes the color and intensity and calls a pair of helper functions, [`term_colour()` and `term_intensity()`](https://github.com/sifive/wake/blob/e596242b4a009550e2bd06ea21206f9aedeeea18/src/runtime/term.cpp#L42-L57).

`term_colour()` is basically just a wrapper around the [`term` `setaf`](https://man7.org/linux/man-pages/man5/terminfo.5.html) capability, which gets the control code for setting the foreground color to an ANSI color.

`term_intensity()` looks up the control code for setting dim or bold (bright). The actual behavior of dim or bold depends on your terminal emulator (some like mine _actually_ bold the text, but some will just change the text to a brighter color).

Anyway, based on my reading of this code, I think you are right. Dim will send the Dim control code to the user's terminal. Most non-Scala programs will not attempt to set color codes when run under wake, but Scala programs will, and setting the color intensity to dim will actually dim the text if the program set color codes (unless the program overrides the dimness code).

Reply:
Sorry, I need to remember to refresh before hitting submit, especially when I've been typing comments for a long time. Didn't see your previous comment.

> the program's original color codes come out. I mean, there are 2 arguments to the color (in wake) -- the color override and the intensity override. the Dim color used by logInfo only overrides the second one without touching the first one. "console default" isn't the right term, but I want to say something like "whatever color you would have seen if you were just running it, not using wake". do you have a better suggestion for that term?

I think we can just leave it at dim. I wasn't sure if wake was itself shifting into 4-bit ANSI, which has a "bright" set of the original 8 colors in 3-bit ANSI, but it looks like what is labeled as "bright" in wake actually corresponds to "bold" in ANSI, not "bright" in ANSI, so I was confused.

I would really suggest renaming the Wake enum value from `Bright` to `Bold`, since not everyone's terminal emulators will actually brighten the text (mine will in fact bold it).

Reply:
ok, i think renaming `Bright` to `Bold` is outside the scope of this doc, and I think this thread is mostly resolved...? I decided to keep the colors in the existing table but made it focus on the fact that these are the loggers, because the colors are intrinsic to the loggers